src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/TagHelperCompletionProvider.cs

```diff
@@ -58,28 +58,27 @@ public ImmutableArray<RazorCompletionItem> GetCompletionItems(RazorCompletionCon
             return ImmutableArray<RazorCompletionItem>.Empty;
         }
 
-        var parent = owner.Parent;
-
-        // When overtyping a self closing tag completion happens on "<c$$ />" and the owner is a "misc attribute content"
-        // so we have to navigate back up to the start tag to provide completion. Need to be careful though as we don't
-        // want element completions for <c $$ /> even though the parent node is the same
-        if (parent is MarkupMiscAttributeContentSyntax { Parent: MarkupStartTagSyntax startTag } &&
-            context.AbsoluteIndex == startTag.Name.Span.End)
+        owner = owner switch
         {
-            parent = startTag;
-        }
+            MarkupStartTagSyntax or MarkupEndTagSyntax or MarkupTagHelperStartTagSyntax or MarkupTagHelperEndTagSyntax => owner,
```

Comment:
I can't remember why this came up as needing special casing now. If you can, perhaps a comment above would be helpful?

Reply:
Will add my best understanding.